spec.bs

```diff
@@ -221,6 +226,55 @@ The <dfn for="interest group">estimated size</dfn> of an [=interest group=] |ig|
     [=interest group ad/render url=].
   1. The [=string/length=] of |ad|'s [=interest group ad/metadata=] if the field is not null.
 
+<div algorithm>
+
+To <dfn>check interest group permissions</dfn> given an [=url/origin=]
+|ownerOrigin|, an [=url/origin=] |frameOrigin|, and a [=boolean=] |isJoin|:
+1. If |ownerOrigin| is [=same origin=] to |frameOrigin|, then return true.
+1. Let |permissionsUrl| be the result of [=building an interest group permissions url=]
+  with |ownerOrigin| and |frameOrigin|.
+1. Let |resource| be the result of [=fetching resource=] with |permissionsUrl|
+  and "`application/json`".
+1. Let |request| be the result of [=creating a request=] with |permissionsUrl|, [=create a request/accept=]
+  set to "`application/json`"| and [=create a request/corsMode=] set to true.
+1. Let |resource| be null.
+1. [=Fetch=] |request| with [=fetch/processResponseConsumeBody=] set to the following steps given a
+  [=response=] |response| and |responseBody|:
+  1. If |responseBody| is null or failure, set |resource| to failure and return.
+  1. Let |headers| be |response|'s [=response/header list=].
+  1. If [=header list/getting a structured field value=] "X-Allow-FLEDGE" from |headers| does not
+    return true, set |resource| to failure and return.
+  1. Let |headerMimeType| be the result of [=header list/extracting a MIME type=] from |headers|.
+  1. Set |resource| to failure and return, if any of:
+    1. |headerMimeType| is failure.
+    1. |headerMimeType| is not a [=JSON MIME type=].
+  1. Let |mimeTypeCharset| be |headerMimeType|'s [=MIME type/parameters=]["`charset`"].
+  1. Set |resource| to failure and return if any of:
+    1. If |mimeTypeCharset| does not [=map/exist=], or |mimeTypeCharset| is "utf-8", and
+      |responseBody| is not [=UTF-8=] encoded.
+    1. If |mimeTypeCharset| is "us-ascii", and |responseBody| is not [=ascii string=].
+  1. Set |resource| to |responseBody|.
+1. Wait for |resource| to be set.
+1. If |resource| is failure, then return false.
+1. Let |permissions| be the result of [=parsing a JSON string to an Infra value=], returning false on failure.
+1. If |isJoin| is true and |permissions| is a JSON object containing key
+  "`joinAdInterestGroup`" with value `true`, return true.
+1. If |isJoin| is false and |permissions| is a JSON object containing key
+  "`leaveAdInterestGroup`" with value `true`, return true.
+1. Return false.
+
+</div>
+
+<div algorithm>
+
+To <dfn>build an interest group permissions url</dfn> given a [=url/origin=] |ownerOrigin| and a [=url/origin=] |frameOrigin|:
+1. Return the string formed by [=string/concatenating=]
+  * The [=serialization of an origin|serialization=] of |ownerOrigin|,
+  * The string "`/.well-known/interest-group/permissions/?origin=`", and
+  * The [=serialization of an origin|serialization=] of |frameOrigin|.

```

Comment:
We need to encode |frameOrigin| when using it in the URL's query param. So maybe something like:
Add this before the step:
`Let |serializedFrameOrigin| be the result of [ =serialization of an origin|serializing=] |frameOrigin|.`
Then here:
`* The result of [=string/UTF-8 percent-encoding=] |serializedFrameOrigin| using [=component percent-encode set=].`